src/libraries/System.Private.Xml/src/System/Xml/BinHexDecoder.cs

```diff
@@ -180,22 +160,23 @@ public static unsafe byte[] Decode(char[] chars, bool allowOddChars)
         // Private methods
         //
 
-        private static unsafe void Decode(char* pChars, char* pCharsEndPos,
-                                    byte* pBytes, byte* pBytesEndPos,
-                                    ref bool hasHalfByteCached, ref byte cachedHalfByte,
-                                    out int charsDecoded, out int bytesDecoded)
+        private static void Decode(ReadOnlySpan<char> chars,
+                                   Span<byte> bytes,
+                                   ref bool hasHalfByteCached, ref byte cachedHalfByte,
+                                   out int charsDecoded, out int bytesDecoded)
         {
-#if DEBUG
-            Debug.Assert(pCharsEndPos - pChars >= 0);
-            Debug.Assert(pBytesEndPos - pBytes >= 0);
-#endif
+            int iByte = 0;
+            int iChar = 0;
 
-            char* pChar = pChars;
-            byte* pByte = pBytes;
-            while (pChar < pCharsEndPos && pByte < pBytesEndPos)
+            for (; iChar < chars.Length; iChar++)
             {
+                if ((uint)iByte >= (uint)bytes.Length)
```

Comment:
Code is written this way to avoid bounds checks later in the method.

Reply:
@GrabYourPitchforks, (maybe you already know it but) since https://github.com/dotnet/runtime/pull/40180 was merged, the `uint` cast workaround is not needed in some cases where constant operands are involved. It turned out that in some cases, RyuJIT does not elide the bound check with `uint` casts, where it was doing before. I posted some findings on macOS here: https://github.com/dotnet/runtime/issues/11623#issuecomment-704607136. It is unlikely that this construct with (non const) operands is affected by that change, but perhaps would be good to double check with the latest master. I have a feeling that in some places in the framework, we can remove the `uint` cast, as they are deoptomized.

Reply:
@am11 This is a good observation! As these other optimizations come online I think it'll be useful to perform a libraries-wide sweep of all of these patterns. It's always great to make the code more readable while maintaining peak efficiency. :)